File: client/lib/api/v2/schemas/practice.ts
/**
 * Practice Overview Schemas
 */

import { z } from "zod";
import {
  ParameterItemMappingSchema,
  ParameterMappingSchema,
  PersonaMappingSchema,
  ScenarioMappingSchema,
  SimulationMappingSchema,
  StandardGroupsMappingSchema,
  StandardsMappingSchema,
} from "./base";
import { AttemptHistoryResponseSchema } from "./home";

/**
 * Practice filter schema - simplified to profile-only
 */
export const PracticeFiltersSchema = z.object({
  profileId: z.string(),
  departmentIds: z.array(z.string()).optional(),
});

export type PracticeFilters = z.infer<typeof PracticeFiltersSchema>;

export const PracticeSimulationItemSchema = z.object({
  viewMode: z.enum(["practice"]),
  id: z.string(),
  simulationTitle: z.string(),
  simulationDescription: z.string().nullable(),
  simulationName: z.string(),
  timeLimit: z.number().nullable().optional(),
  numSessions: z.number(),
  highestScore: z.number().nullable().optional(),
  standard_groups: z.record(z.string(), z.array(z.string())),
  rubric_id: z.string().nullable().optional(),
  color: z.string().nullable().optional(),
  icon: z.string().nullable().optional(),
  hasPassed: z.boolean().nullable().optional(),
  passRate: z.number().nullable().optional(),
  status: z
    .enum(["not-started", "in-progress", "passed"])
    .nullable()
    .optional(),
  completionPct: z.number().nullable().optional(),
  passedCount: z.number().nullable().optional(),
  inProgressCount: z.number().nullable().optional(),
  notStartedCount: z.number().nullable().optional(),
  passPct: z.number().nullable().optional(),
  cohortName: z.string().nullable().optional(),
  updatedAt: z.string().nullable().optional(),
  lastActivityTs: z.string().nullable().optional(),
  hasActivity: z.boolean().nullable().optional(),
});

export const PracticeOverviewResponseSchema = z.object({
  mode: z.enum(["practice"]),
  hasData: z.boolean(),
  items: z.array(PracticeSimulationItemSchema),
  history: AttemptHistoryResponseSchema,
  standard_groups_mapping: StandardGroupsMappingSchema,
  standards_mapping: StandardsMappingSchema,
  simulation_mapping: SimulationMappingSchema,
  persona_mapping: PersonaMappingSchema,
  scenario_mapping: ScenarioMappingSchema,
  parameter_mapping: ParameterMappingSchema,
  parameter_item_mapping: ParameterItemMappingSchema,
});

export type PracticeSimulationItem = z.infer<
  typeof PracticeSimulationItemSchema
>;
export type PracticeOverviewResponse = z.infer<
  typeof PracticeOverviewResponseSchema
>;

File: server/app/schemas/practice.py
"""Practice overview schemas."""

from typing import Literal

from pydantic import BaseModel

from .base import (ParameterItemMapping, ParameterMapping, PersonaMapping,
                   ScenarioMapping, SimulationMapping, StandardGroupsMapping,
                   StandardsMapping)
from .home import AttemptHistoryResponse


class PracticeFilters(BaseModel):
    """Practice filter request schema - simplified to profile-only."""

    profileId: str
    departmentIds: list[str] | None = None


class PracticeSimulationItem(BaseModel):
    """Practice simulation item."""

    viewMode: Literal["practice"]
    id: str
    simulationTitle: str
    simulationDescription: str | None = None
    simulationName: str
    timeLimit: int | None = None
    numSessions: int
    highestScore: float | None = None
    standard_groups: dict[str, list[str]]
    color: str | None = None
    icon: str | None = None
    hasPassed: bool | None = None
    passRate: float | None = None
    status: Literal["not-started", "in-progress", "passed"] | None = None
    completionPct: float | None = None
    passedCount: int | None = None
    inProgressCount: int | None = None
    notStartedCount: int | None = None
    passPct: float | None = None
    cohortName: str | None = None
    updatedAt: str | None = None
    lastActivityTs: str | None = None
    hasActivity: bool | None = None


class PracticeOverviewResponse(BaseModel):
    """Practice overview response with mappings and history."""

    mode: Literal["practice"]
    hasData: bool
    items: list[PracticeSimulationItem]
    history: AttemptHistoryResponse
    standard_groups_mapping: StandardGroupsMapping
    standards_mapping: StandardsMapping
    simulation_mapping: SimulationMapping
    scenario_mapping: ScenarioMapping
    persona_mapping: PersonaMapping
    parameter_mapping: ParameterMapping
    parameter_item_mapping: ParameterItemMapping
